src/replication/ng/registry/manager.go

```diff
@@ -15,20 +15,326 @@
 package registry
 
 import (
+	"fmt"
+	"net/http"
+
+	"github.com/goharbor/harbor/src/common/dao"
+	"github.com/goharbor/harbor/src/common/models"
+	"github.com/goharbor/harbor/src/common/utils"
+	utilerr "github.com/goharbor/harbor/src/common/utils/error"
+	"github.com/goharbor/harbor/src/common/utils/log"
+	"github.com/goharbor/harbor/src/common/utils/registry"
+	"github.com/goharbor/harbor/src/common/utils/registry/auth"
+	"github.com/goharbor/harbor/src/core/config"
 	"github.com/goharbor/harbor/src/replication/ng/model"
 )
 
-// Manager manages registries
+// HealthStatus describes whether a target is healthy or not
+type HealthStatus string
+
+const (
+	// Healthy indicates target is healthy
+	Healthy = "healthy"
+	// Unhealthy indicates target is unhealthy
+	Unhealthy = "unhealthy"
+	// Unknown indicates health status of target is unknown
+	Unknown = "unknown"
+)
+
+// Manager defines the methods that a target manager should implement
 type Manager interface {
 	// Add new registry
 	Add(*model.Registry) (int64, error)
 	// List registries, returns total count, registry list and error
 	List(...*model.RegistryQuery) (int64, []*model.Registry, error)
 	// Get the specified registry
 	Get(int64) (*model.Registry, error)
+	// GetByName gets registry by name
+	GetByName(name string) (*model.Registry, error)
+	// GetByURL gets registry by its URL
+	GetByURL(url string) (*model.Registry, error)
 	// Update the registry, the "props" are the properties of registry
 	// that need to be updated
 	Update(registry *model.Registry, props ...string) error
 	// Remove the registry with the specified ID
 	Remove(int64) error
+	// HealthCheck checks health status of all registries and update result in database
+	HealthCheck() error
+}
+
+// DefaultManager implement the Manager interface
+type DefaultManager struct{}
+
+// NewDefaultManager returns an instance of DefaultManger
+func NewDefaultManager() *DefaultManager {
+	return &DefaultManager{}
+}
+
+// Ensure *DefaultManager has implemented Manager interface.
+var _ Manager = (*DefaultManager)(nil)
+
+// Get gets a registry by id
+func (m *DefaultManager) Get(id int64) (*model.Registry, error) {
+	registry, err := dao.GetRegistry(id)
+	if err != nil {
+		return nil, err
+	}
+
+	if registry == nil {
+		return nil, utilerr.KnownError{
```

Comment:
This isn't consistent with the logic in `GetByName`, I think the logic in `GetByName` is better, it's easy for API level to handle 404.